src/gpu/align_z_to_other.cu

```diff
@@ -0,0 +1,239 @@
+            return stream;
+        }
+
+
+    };
+    static std::mutex gpu_mutex;
+    __host__ bool align_z_to_other(rsimpl::byte * z_aligned_to_other, const uint16_t * z_pixels, float z_scale, const rs_intrinsics & z_intrin, const rs_extrinsics & z_to_other, const rs_intrinsics & other_intrin)
+    {
+        DEBUG("GPU NOT DISABLED");
+
+        int deviceCount = 0;
+        auto cudaCallErrorStatus = cudaGetDeviceCount(&deviceCount);
+        if (cudaSuccess != cudaCallErrorStatus) {
+            std::cerr << "Failed to obtain number of GPUs. " << cudaGetErrorName(cudaCallErrorStatus) << " " << cudaGetErrorString(cudaCallErrorStatus) << std::endl;
+            return false;
+        }
+        if (deviceCount == 0) {
+            DEBUG("NO ENABLED GPUs PRESENT");
+            return false;
+        }
+
+
+        // ignore this error. technically the blocking sync can only be set once prior to device initialization. once
+        // that initial setting is done, subsequent settings result in an error until device is restarted
+        unsigned int cudaFlags = 0;
+        cudaCallErrorStatus = cudaGetDeviceFlags(&cudaFlags);
+        DEBUG("Flags is set to: " << cudaFlags);
+        DEBUG("Error status is: " << cudaCallErrorStatus);
+        if (!(cudaFlags & cudaDeviceScheduleBlockingSync)) {
+            DEBUG("attempting to set the cudaDeviceScheduleBlockingSync flag");
+            std::lock_guard<std::mutex> lock(gpu_mutex);
+            // check flag again to make sure we don't reset it after some other thread set it for us
+            cudaCallErrorStatus = cudaGetDeviceFlags(&cudaFlags);
+            if (!(cudaFlags & cudaDeviceScheduleBlockingSync)) {
+                DEBUG("flag still not set");
+                cudaCallErrorStatus = cudaSetDeviceFlags(cudaDeviceScheduleBlockingSync);
+                if (cudaCallErrorStatus != cudaSuccess) {
+                    DEBUG("reseting device");
+                    cudaDeviceSynchronize();
+                    cudaDeviceReset();
+                    cudaCallErrorStatus = cudaSetDeviceFlags(cudaDeviceScheduleBlockingSync);
+                }
+            }
+        }
+
+
+        auto out_z = (uint16_t *)(z_aligned_to_other);
+
+        auto cudaDeleterUnsignedInt = [](unsigned int *ptr) {cudaFree(ptr);};
+        auto cudaDeleterUInt16t = [](uint16_t *ptr) {cudaFree(ptr);};
+        CudaStreamWrapper myCudaStream;
+        // Kernel invocation
+        dim3 threadsPerBlock(32, 32);
+        dim3 numBlocks(static_cast<int> (std::ceil(static_cast<float>(z_intrin.width) / threadsPerBlock.x)),
+                       static_cast<int>(std::ceil(
+                static_cast<float>(z_intrin.height) / threadsPerBlock.y)));
+        uint16_t * z_pixels_gpu;
+        unsigned int * temporary_out_pixels_gpu;
+        uint16_t * out_pixels_gpu_uint16t;
+
+
+        cudaCallErrorStatus = cudaMalloc((void **)&z_pixels_gpu, sizeof(uint16_t) * z_intrin.width * z_intrin.height);
```

Comment:
Where do we free the GPU memory?

Reply:
Can we also allocate in the constructor?  Might save some more time, since we will always be allocating/freeing memory.

Reply:
each malloc is followed by:
std::unique_ptr<uint16_t, decltype(cudaDeleterUInt16t)> z_pixels_gpu_u_ptr(z_pixels_gpu, cudaDeleterUInt16t);